Shrink `DatabaseKeyIndex` to 8 bytes to save memory

ChayimFriedman2 opened the pull request:
Via the observation that:

 - The `IngredientIndex` can be retrieved from the page, except for tracked fns
 - The generation is useless, except for interneds

And since they don't overlap, we can store only what needed for each.

For rust-analyzer, this saves 55mb on itself and 200mb on a large codebase (omicron).

ChayimFriedman2 committed:
Shrink `DatabaseKeyIndex` to 8 bytes to save memory

Via the observation that:

 - The `IngredientIndex` can be retrieved from the page, except for tracked fns
 - The generation is useless, except for interneds

And since they don't overlap, we can store only what needed for each.

MichaReiser commented:
Nice, there are some test failures (re-running fixed the temporary out-of-disk-space error)

Veykril commented:
Okay I deleted the ci caches, lets see if that helps

Veykril commented:
> Unfortunately, this breaks our test infrastructure:

We can probably change the fields of that event and just deconstruct the key eagerly, since the event infra is callback based one only pays when a callback is registered after all

Veykril commented:
Or alternatively expose a public getter that takes a `&dyn Database` instead

ChayimFriedman2 committed:
Use a generationless ID in the sync table

Because we sometimes get id from a `DatabaseKeyIndex` that doesn't have the generation, so it prevent conflicts. And we don't need it either, as explained in a comment.

ibraheemdev commented on src/tracked_struct.rs:
```diff
@@ -480,9 +476,14 @@ where
     /// Returns the database key index for a tracked struct with the given id.
     pub fn database_key_index(&self, id: Id) -> DatabaseKeyIndex {
-        DatabaseKeyIndex::new(self.ingredient_index, id)
+        DatabaseKeyIndex::new_non_interned(self.ingredient_index, id)
```
We use generations for tracked structs that are reused to avoid an extra read dependency, not just interned values. This might be the cause of the test failures?

MichaReiser commented:
We now store a compact 8 byte `DatabaseKeyIndex` in `QueryEdges` for the common case (reasonable generation, reasonable ingredient index). There are a few other places where we store `DatabaseKeyIndex`, but they aren't as common, where storing 12 bytes is probably fine. That's why I'd suggest closing this PR for now and revisiting this optimization on a concrete use case.
